split recording session events and emitting audit events

capnspacehook opened the pull request:
**NOTE** many of the changes are renames or small fixes to existing tests.

This is a refactor of how audit events and session events are handled. Previously, all events were emitted using the same interface, `api/types/events.Emitter`. This lead to event-related code getting to be very confusing, as it was often unclear whether a given event was being recorded as a session event *and* emitted as an audit event, or only one of the two. Naturally, a few bugs arose due to this.

Using the comments in   as a guide, a separate interface for recording session events has been created. A `api/types/events.Recorder` should *only* be used to record session events, and an `Emitter` should now only be used to emit audit events. Instead of using a confusing `TeeWriter` that would transparently (and confusingly, given its name) hold a few event types that only belonged in session recordings, callers can now explicitly record and/or emit an event when necessary.

---

Manual testing I preformed:

For each item in the following list, I ensured events were recorded and/or emitted as necessary

- SSH:
  - Connect to node
  - Join/leave session
- Kube:
  - Connect to kube
- App Access:
  - Make requests to web app
- DB:
  - Login to database
  - Make queries to db
- Desktop Access:
  - Establish RDP session with a host

Merging this will break builds to `e`, but I'll have a small PR that will fix `e` builds ready before I merge this.

Fixes https://github.com/gravitational/teleport/issues/16773.
Fixes https://github.com/gravitational/teleport/issues/25115.

zmb3 commented on integration/integration_test.go:
```diff
@@ -4543,14 +4562,20 @@ func testAuditOff(t *testing.T, suite *integrationTestSuite) {
+	sessionsDir := t.TempDir()
 	makeConfig := func() (*testing.T, []string, []*helpers.InstanceSecrets, *servicecfg.Config) {
+		auditConfig, err := types.NewClusterAuditConfig(types.ClusterAuditConfigSpecV2{
+			AuditSessionsURI: sessionsDir,
```
I'm surprised this works. Doesn't `AuditSessionsURI` have to be a URI? (`file:///home/log`)

capnspacehook replied:
I guess the scheme defaults to `file` if it's unset: https://github.com/gravitational/teleport/blob/capnspacehook/seperate-audit-record-events/api/utils/uri.go#L35

zmb3 commented on lib/auth/helpers.go:
```diff
@@ -68,8 +69,10 @@ type TestAuthServerConfig struct {
 	// networking configuration.
 	ClusterNetworkingConfig types.ClusterNetworkingConfig
-	// Streamer allows a test to set its own audit events streamer.
+	// Streamer allows a test to set its own session events streamer.
```
```suggestion
	// Streamer allows a test to set its own session recording streamer.
```

zmb3 commented on lib/events/api.go:
```diff
@@ -799,10 +799,14 @@ type UploadMetadataGetter interface {
+// SessionRecorder records session events. It can be used both as a
+// [io.Writer] when recording raw session data and as a [apievents.Recorder]
+// when recording session events.
+type SessionRecorder interface {
```
So we have both an `events.Recorder` interface and an `events.SessionRecorder` interface in this package, and they both are used for session recording? That's a bit confusing IMO.

capnspacehook replied:
Yeah that's a good point, I don't need both those interfaces now. Removed `Recorder` and `StreamManager`.

zmb3 commented on lib/events/emitter.go:
```diff
@@ -614,13 +572,13 @@ func (s *CallbackStream) Complete(ctx context.Context) error {
 }
 
 // EmitAuditEvent emits audit event
```
Fix godoc here

zmb3 commented on lib/events/eventstest/mock_emitter.go:
```diff
@@ -30,6 +30,14 @@ type MockEmitter struct {
 	events []events.AuditEvent
 }
 
+func (e *MockEmitter) Write(_ []byte) (int, error) {
```
```suggestion
func (e *MockEmitter) Write(b []byte) (int, error) {
```

Or `_` out `event` below. Just be consistent.

zmb3 commented on lib/events/recorder/recorder.go:
```diff
@@ -0,0 +1,58 @@
+// recording is set to be synchronous, the returned recorder will use
+// syncStream to create an event stream. Otherwise, a streamer will be
+// used that will back recorded session events to disk for eventual upload.
+func NewRecorder(recCfg types.SessionRecordingConfig, cfg events.SessionWriterConfig, uploadDir string, syncStream events.Streamer) (events.SessionRecorder, error) {
```
Hmm, this feels like a broken abstraction. There's 3 possible outcomes here in the happy path:

1. we produce a discard stream because recording is disabled.
2. we are doing async recording to disk
3. we are streaming in sync mode

We have to pass an upload dir, even though it's only used in 1 of 3 scenarios, and we have to pass a sync streamer even though it's only used in 1 of 3 scenarios.

At this point, what good is the `NewRecorder` helper func if you have to create everything ahead of time and pass it in?

capnspacehook replied:
I see what you mean, it's not a perfect solution by any means, but all services that record sessions duplicated this exact logic before. My thought was to centralize it somewhere to prevent new services from configuring a session recorder incorrectly, as even with all these changes handling events it not super straightforward.

zmb3 commented on lib/events/session_writer.go:
```diff
@@ -76,15 +77,12 @@ type AuditWriterConfig struct {
-	// For backwards compatibility, AuditWriter will convert bytes to
+	// For backwards compatibility, SessionWriter will convert bytes to
 	// SessionPrint events when MakeEvents is not provided.
 	MakeEvents func([]byte) []apievents.AuditEvent
```
Some of the bugs we had with things not getting recorded correctly were due to us mixing the `io.Writer` adapter in with a general purpose session recorder.

I wonder if this would be simpler if `SessionWriter` didn't implement the `io.Writer` interface and only exposed the generic `RecordEvent` method. Then we could have a separate adapter for cases where we need `io.Writer`.

I'm thinking something like:

```
// ByteStreamRecorder records an arbitrary stream of bytes by converting
// them to SessionPrint events.
type ByteStreamRecorder struct {
    recorder events.Recorder
}

func (b *ByteStreamRecorder) Write(b []byte) (n int, err error) {
    // package b up into one or more SessionPrint events
    b.recorder.RecordEvent(printEvent)
}
```

capnspacehook replied:
Hmm, that's not a bad idea, however I think this could be made more simple by returning different interfaces from the `recorder` package. One function could return a session recorder interface without `io.Writer`, and another function could wrap it and return an interface that includes `io.Writer`.

That would avoid creating another arbitrary type and keep all session event logic in SessionWriter, while also ensuring callers only get access to functionality that they need.
```suggestion
	MakeEvents func([]byte) []apievents.AuditEvent
```

capnspacehook replied:
After the changes to separate event setup and recording, doing this would necessitate adding 2 or more interfaces, a `SessionRecorder` without `io.Writer`, and `SessionSetterRecorder` without `io.Writer`. Not sure if this is worth making the event interface count even larger.

zmb3 commented on lib/events/session_writer.go:
```diff
@@ -375,11 +370,33 @@ func (a *AuditWriter) EmitAuditEvent(ctx context.Context, event apievents.AuditE
+		return trace.Wrap(err)
+	}
+	if event.GetTime().IsZero() {
+		return trace.BadParameter("Time is unset, call SetupEvent before recording events")
```
This is not a good message for a user - it doesn't tell them what's gone wrong or how to fix it.

capnspacehook replied:
I agree, though my goal for these errors is to let devs know when they forgot to call `SetupEvent` first. We really need some kind of debug-panic log or something, so we can panic and let developers know when they're not using something correctly without panicking when run in production.

zmb3 commented on lib/events/api.go:
```diff
@@ -799,10 +799,14 @@ type UploadMetadataGetter interface {
 	io.Writer
+	// SetupEvent will set necessary event fields, and must be called before
+	// the event is recorded.
+	SetupEvent(event apievents.AuditEvent) error
```
Does this have to be in the recorder interface?

With this, each recorder, whether it be a streaming recorder, one that writes to disk, a test one that just saves a copy in memory, or any future implementation is forced to implement the same logic for setting necessary fields.

Can we clean this up by either:
- moving the event setup upstream, so the recorder just records, and whatever calls the recorder is responsible for providing valid data
- wrapping with a checking streamer (or whatever its called now) in order to centralize this logic

capnspacehook replied:
The way it's implemented now, `SessionWriter` sets event indexes which requires keeping state. This code could be split from `SessionWriter`, but as we only have 2 recorders now that we have been exclusively for years (`SessionWriter` and `DiscardRecorder`), I think it's better to cross this bridge if we get to it.

capnspacehook replied:
The reason I split `SetupEvent` is because previously, `lib/events.SessionWriter` modified events in the caller's goroutine (called SetupEvent) and `lib/events.AsyncEmitter` read the same fields of the same events in a different goroutine, so events had to be recorded before they were emitted if doing both was desired to prevent races. These 2 structs are always used to record and emit respectively in production.

So centralizing this logic would be nice, but would open us up to easily adding races in the future by accident.

zmb3 replied:
I still think we can fix the race without requiring a bunch of independent implementations of `SetupEvent`. Let me know if you want to brainstorm some solutions here,.

capnspacehook replied:
Updated to have a separate interface for setting up events.

zmb3 commented on lib/srv/app/server_test.go:
```diff
@@ -921,19 +915,19 @@ func TestRequestAuditEvents(t *testing.T) {
 		// wait until chunk events are generated before closing the server.
 		require.Eventually(t, func() bool {
-			return chunkEventsReceived.Load() == 1
+			return chunkEventsReceived.Load() > 0
```
What's the reasoning behind this change?

zmb3 commented on lib/srv/desktop/tdp/proto.go:
```diff
@@ -1469,8 +1467,10 @@ const tdpMaxNotificationMessageLength = 10240
+const (
+	maxClipboardDataLength    = 1024 * 1024 // 1MB
+	tdpMaxFileReadWriteLength = 1024 * 1024 // 1MB
+)
```
Guessing this was an auto-generated change?

We have several consts here, and two of them were combined. I'd revert the change or put all of the consts together.  Grouping some but not all of them makes it seems like these 2 values are in some way related, and I don't think that's actually the case.

capnspacehook replied:
Yep that was a change from a formatter, I'll group all the consts in the same block

zmb3 commented on lib/events/recorder/recorder.go:
```diff
@@ -0,0 +1,58 @@
+// recording is set to be synchronous, the returned recorder will use
+// syncStream to create an event stream. Otherwise, a streamer will be
+// used that will back recorded session events to disk for eventual upload.
+func NewRecorder(recCfg types.SessionRecordingConfig, cfg events.SessionWriterConfig, uploadDir string, syncStream events.Streamer) (events.SessionRecorder, error) {
```
The package is already called `recorder`, so how about `recorder.New` for a name?

capnspacehook replied:
Good idea, done

zmb3 commented on lib/srv/sess.go:
```diff
@@ -922,6 +928,9 @@ func (s *session) emitSessionJoinEvent(ctx *ServerContext) {
 	}
 
 	// Emit session join event to Audit Log.
```
The comment now refers to the line below. I would move it down, or just remove it.

zmb3 commented on lib/srv/sess.go:
```diff
@@ -1240,43 +1255,42 @@ func (s *session) startTerminal(ctx context.Context, scx *ServerContext) error {
+	}
+	// Return an SessionWriter with a discarding stream so closing it will
+	// work correctly for tests
+	if ctx.IsTestStub {
```
I know you didn't write this, but I do wonder if any of your changes would make it easier to get rid of this. Having any sort of `if isTest` logic in production code is a smell.

capnspacehook replied:
Good point, I'll try and remove it

zmb3 reviewed (commented):
Definitely an improvement! I'm happy to see the confusing `TeeWriter` removed.

I'll give this a try tomorrow and see if it fixes some of the recording issues that I've run into.

I still find the sheer amount of types with similar names confusing. 

We have:

- Recorder
- Stream
- StreamManager
- SessionRecorder
- SessionWriter

These all kind of start to sound the same after a while.

greedy52 commented on lib/srv/app/session.go:
```diff
@@ -286,51 +288,25 @@ func (s *Server) newStreamWriter(chunkID string) (events.StreamWriter, error) {
 		events.StreamingSessionsDir, apidefaults.Namespace,
 	)
-	fileStreamer, err := filesessions.NewStreamer(uploadDir)
+	streamWriter, err := recorder.New(recConfig, cfg, uploadDir, s.c.AuthClient)
```
previous implementation using  `events.NewTeeStreamer` avoids sending `AppSessionRequestEvent` events to audit event (`AppSessionRequestEvent` only goes to session). I believe that logic now should be adjusted in `common.Audit`

capnspacehook replied:
Thanks and good catch, fixed

capnspacehook commented:
@zmb3 @greedy52 ready for another look

r0mant commented:
/excludeflake *

greedy52 commented on lib/events/recorder/recorder.go:
```diff
@@ -0,0 +1,73 @@
+	}
+	swCfg.Streamer = streamer
+
+	rec, err := events.NewSessionWriter(swCfg)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
```
curious if SessionWriter can/should implement events.SessionSetterRecorder as it owns a Setting from config anyway.

capnspacehook replied:
Was on the fence about that, but I think it would help if it did. Pushed a commit that adds this

greedy52 commented on lib/srv/desktop/windows_server.go:
```diff
@@ -952,33 +919,40 @@ func (s *WindowsService) makeTDPSendHandler(ctx context.Context, emitter events.
-			} else if err := emitter.EmitAuditEvent(ctx, e); err != nil {
-				s.cfg.Log.WithError(err).Warning("could not emit desktop recording event")
+			} else {
+				event, err := recorder.SetupSessionEvent(e)
+				if err == nil {
+					if err := recorder.RecordEvent(ctx, event); err != nil {
+						s.cfg.Log.WithError(err).Warning("could not record desktop recording event")
```
nit: not sure if worth it to have a `recorder.SetupAndRecordEvent` as this repeat a few places.

capnspacehook replied:
I think in places where the event will only be recorded this makes sense, done

greedy52 commented on lib/events/session_writer.go:
```diff
@@ -281,17 +259,21 @@ func (a *AuditWriter) maybeSetBackoff(backoffUntil time.Time) bool {
 		return trace.Wrap(err)
 	}
 
+	// the index starts at 0, so we'll only know if the index is invalid
+	// if we've seen at least one event already
+	if event.GetIndex() == 0 && a.acceptedEvents.Load() > 0 {
+		return trace.BadParameter("missing mandatory event index field")
+	}
```
I assume no op setter will never be used with SessionWriter?

capnspacehook replied:
Yep a no-op setter should only be used for tests or when events will be discarded

greedy52 commented on lib/srv/sess.go:
```diff
@@ -919,7 +925,9 @@ func (s *session) emitSessionJoinEvent(ctx *ServerContext) {
 	}
 
-	// Emit session join event to Audit Log.
+	if err := s.recordEvent(ctx.srv.Context(), sessionJoinEvent); err != nil {
+		s.log.WithError(err).Warn("Failed to record session join event.")
+	}
```
i am assuming the reason `s.recordEvent` is before `s.emitAuditEvent` because recordEvent does the setup?

capnspacehook replied:
Yes exactly

capnspacehook commented on lib/events/setter.go:
```diff
@@ -0,0 +1,133 @@
+	}
+
+	event.SetIndex(int64(c.eventIndex.Load()))
+	c.eventIndex.Add(1)
```
Good catch, fixed

capnspacehook commented on lib/events/recorder/recorder.go:
```diff
@@ -0,0 +1,73 @@
+// recording is set to be synchronous, the returned recorder will use
+// syncStream to create an event stream. Otherwise, a streamer will be
+// used that will back recorded session events to disk for eventual upload.
+func New(setterCfg events.SetterConfig, swCfg events.SessionWriterConfig, recCfg types.SessionRecordingConfig, uploadDir string, syncStream events.Streamer) (events.SessionSetterRecorder, error) {
```
`New` doesn't do too much that is true, but it does bundle the exact same (somewhat non-obvious) code used almost half a dozen times in the codebase into one place, making it easy to update in the future. 5 or so repetitions was enough to make me decide to put it into a function.

zmb3 commented on lib/events/api.go:
```diff
@@ -799,13 +799,28 @@ type UploadMetadataGetter interface {
 
+// SessionSetterRecorder sets necessary session event fields and records them.
+type SessionSetterRecorder interface {
+	SetupAndRecordEvent(ctx context.Context, event apievents.AuditEvent) error
```
If a concrete type implements both `SessionEventSetter` and `SessionRecorder` doesn't it have everything it needs?

(Why do we need to require that it also implements `SetupAndRecordEvent`?)

capnspacehook replied:
This was a bit of a smell, making `SetupAndRecordEvent` a function is a much better idea

zmb3 commented on lib/events/api.go:
```diff
@@ -799,13 +799,28 @@ type UploadMetadataGetter interface {
+	SessionRecorder
+}
+
 // StreamEmitter supports submitting single events and streaming
 // session events
 type StreamEmitter interface {
```
```suggestion
// StreamEmitter supports emitting single events to the audit log
// and streaming events to a session recording.
type StreamEmitter interface {
```

zmb3 commented on lib/events/discard.go:
```diff
@@ -64,38 +66,60 @@ func (d *DiscardAuditLog) StreamSessionEvents(ctx context.Context, sessionID ses
-func (*DiscardStream) Complete(ctx context.Context) error {
+// Complete marks the stream as closed
+func (d *DiscardRecorder) Complete(ctx context.Context) error {
+	if !d.completed.Load() {
+		close(d.done)
+	}
+	d.completed.Store(true)
```
You'll want to use a compare and swap here, otherwise it's possible that the close call happens twice.
```suggestion
	if !d.completed.CompareAndSwap(false, true) {
		close(d.done)
	}
```

capnspacehook replied:
Thanks, done

zmb3 commented on lib/events/setter.go:
```diff
@@ -0,0 +1,148 @@
+
+// SetupAndRecordEvent will set necessary event fields for session-related
+// events and record them.
+func (s SetterAndRecorder) SetupAndRecordEvent(ctx context.Context, event apievents.AuditEvent) error {
```
So if you removed `SetupAndRecordEvent` from `SessionSetterRecorder`, then this could just be a free function:

```
func SetupAndRecord(ctx context.Context, s SessionSetterRecorder, event apievents.AuditEvent) error {
    // the body remains the same, but now you no longer
    // need a new struct for SetterAndRecorder
}
```

capnspacehook replied:
That's a good idea, done

zmb3 commented on lib/kube/proxy/sess.go:
```diff
@@ -561,10 +564,16 @@ func (s *session) launch() error {
+			s.forwarder.log.WithError(err).Warn("Failed to emit session start event.")
+		}
+	} else {
+		s.forwarder.log.WithError(err).Warn("Failed to setup session start event.")
```
```suggestion
		s.forwarder.log.WithError(err).Warn("Failed to set up session start event - event will not be recorded")
```

zmb3 commented on lib/srv/sess.go:
```diff
@@ -1238,43 +1252,44 @@ func (s *session) startTerminal(ctx context.Context, scx *ServerContext) error {
+			Component: teleport.Component(teleport.ComponentSession, ctx.srv.Component()),
+		},
+		ctx.SessionRecordingConfig,
+		filepath.Join(
+			ctx.srv.GetDataDir(), teleport.LogsDir, teleport.ComponentUpload,
+			events.StreamingSessionsDir, ctx.srv.GetNamespace(),
+		),
```
This logic for building the directory path for a recording seems to be duplicated in a few places - what do you think about a utility in `lib/events` to consolidate this?

capnspacehook replied:
I rolled that into `lib/events/recorder.New` instead and simplified how `recorder.New` is called

zmb3 reviewed (commented):
We're nearly there - a few minor comments while I give this a test drive.

zmb3 commented:
I tried to launch a desktop session and was presented with this error:

<img width="928" alt="image" src="https://github.com/gravitational/teleport/assets/7527103/ff573093-f64b-4bd5-a25e-cc01a9a917be">

capnspacehook commented:
Did another round of manual testing and ensured that error is fixed @zmb3.

capnspacehook commented on lib/events/api.go:
```diff
@@ -799,15 +799,29 @@ type UploadMetadataGetter interface {
+// of whether the event will be recorded, emitted, or both.
+type SessionEventSetter interface {
+	SetupSessionEvent(event apievents.AuditEvent) (apievents.AuditEvent, error)
+}
```
Addressed both suggestions. I had thought before about somehow enforcing that event setup is done before recording but this seems like a clear way to do it. I opted to make `PreparedSessionEvent` a new interface so it can't accidentally be misused by simply casting a `AuditEvent`. `SetupSessionEvent` being separate is a necessary evil unfortunately.

Previously `SessionWriter` would set event fields in the caller's goroutine, and `AsyncEmitter` would check the same event fields in a new goroutine when events were recorded or emitted respectively. It was very easy to trigger a data race by emitting and recording an event in that order. Making event setup separate means the same event can be recorded and emitted in either order with no problems.

capnspacehook commented on integration/integration_test.go:
```diff
@@ -4736,6 +4763,52 @@ func testAuditOff(t *testing.T, suite *integrationTestSuite) {
+	}
+	require.True(t, hasStart)
+	require.True(t, hasLeave)
+	require.True(t, hasEnd)
```
Done

capnspacehook commented:
@nklaassen @zmb3 ready for another look

zmb3 commented on lib/events/setter.go:
```diff
@@ -0,0 +1,169 @@
+	return nil
+}
+
+type Setter struct {
```
missing godocs

zmb3 reviewed (approved):
I can confirm this fixes the desktop recording/playback issue.

capnspacehook commented on lib/events/api.go:
```diff
@@ -799,15 +799,29 @@ type UploadMetadataGetter interface {
+// events and must be called before the event is recorded, regardless
+// of whether the event will be recorded, emitted, or both.
+type SessionEventSetter interface {
+	SetupSessionEvent(event apievents.AuditEvent) (apievents.PreparedSessionEvent, error)
```
Good point, done

capnspacehook commented on api/types/events/api.go:
```diff
@@ -77,11 +77,19 @@ type Emitter interface {
+// a [github.com/gravitational/teleport/lib/events.SessionEventSetter].
+// More specifically, it is a wrapper around an AuditEvent that signifies
+// the event has been prepared and is ready to be recorded or emitted.
+type PreparedSessionEvent interface {
```
I wanted to avoid a exported concrete type so devs couldn't inadvertently create a prepared event without actually preparing it. The way it is now you have to use a preparer from `lib/events` or refactor the existing code so you can create a concrete prepared event yourself which would be a red flag.

zmb3 replied:
Yeah, I agree with that. 

Maybe this is one good use case for an exported method that returns an unexported type.

nklaassen replied:
I may be misremembering but I think our linter will whine about an exported method that returns an unexported type.

I think exporting a concrete type is fine, if a caller is willing to make the explicit cast, they are saying they know what they are doing, they can deal with the consequence. The interface kind of gets in the way, there's at least one place already where you create a nooppreparer just to convert the type, that's a perfect place to just do the type cast I think

zmb3 replied:
Just spitballing here, not even sure I like this idea...

You would have to move some things around - the code that creates prepared events would have to live in api/events instead of lib/events, but then you could have something like:

```
type preparedEvent struct { ... }

func PrepareEvent(event AuditEvent) preparedEvent { ... }

func Record(event preparedEvent) { ... }
```

Normally I'd push back against such a change, as it's kind of a pain to work with public APIs that require unexported types. In this case, it's kind of what you want though: there's only one way to get a `preparedEvent`, so you're forced into doing the right thing and calling `PrepareEvent` first.

If we export `PreparedEvent`, a new dev struggling to get their code to compile might just add an explicit type conversion and move on, not realizing that they broke the rules. (Though maybe Nic's right and this isn't a big deal)

This would also have impacts on your tests, because if you wanted to keep a no-op preparer, it would have to be in the api/events package.

What do you guys think? I'm still not sure, which probably means we shouldn't do it.

capnspacehook replied:
Ah I understand what you mean now. I had the exact same idea, but you're unfortunately right in that it won't work inside `api/events`. Too much code would need `preparedEvent` inside `lib/events` which it wouldn't have access to. I tried moving `Stream` to `lib/events` so `preparedEvent` could live there as well, but that would require `api/events` importing `lib/events` which is a non-starter.

nklaassen reviewed (approved):
I like the `PreparedSessionEvent` changes, makes it clearer to me what you need to call when. I think it may be better as a concrete type, but the interface is alright

capnspacehook committed:
split recording session events and emitting audit events

This is a refactor of how audit events and session events are handled.
Previously, all events were emitted using the same interface,
api/types/events.Emitter. This lead to event-related code getting to be
very confusing, as it was often unclear whether a given event was being
recorded as a session event and emitted as an audit event, or only one
of the two. Naturally, a few bugs arose due to this.

To simplify event handling, a separate interface for recording session
events has been created. A api/types/events.Recorder should now only be
used to record session events, and an Emitter should now only be used to
emit audit events. Instead of using a confusing TeeWriter that would
transparently (and confusingly, given its name) hold a few event types
that only belonged in session recordings, callers can now explicitly
record and/or emit an event when necessary.

codingllama commented on lib/events/eventstest/mock_emitter.go:
```diff
@@ -1,90 +0,0 @@
-)
-
-// MockEmitter is an emitter that stores all emitted events.
-type MockEmitter struct {
```
Should we keep this type as an alias, or just keep the impl itself, while we do the e/ changes? Won't this break the build otherwise?

codingllama replied:
Same for other breaking changes.

capnspacehook replied:
Good point, I'll make an alias

capnspacehook committed:
ensure e build won't break

codingllama commented:
Thanks for putting in the effort to not break e/. Much appreciated!
